aloha-core/src/main/scala/com/eharmony/aloha/dataset/vw/multilabel/VwMultilabelRowCreator.scala

```diff
@@ -0,0 +1,361 @@
+    *                   namespace.
+    * @param classNs a namespace for features associated with class labels
+    * @param dummyClassNs a namespace for features associated with dummy class labels
+    * @return an array to be passed directly to an underlying `VWActionScoresLearner`.
+    */
+  private[aloha] def trainingInput(
+      features: IndexedSeq[Sparse],
+      indices: sci.IndexedSeq[Int],
+      positiveLabelIndices: Int => Boolean,
+      defaultNs: List[Int],
+      namespaces: List[(String, List[Int])],
+      classNs: Char,
+      negativeDummyStr: String,
+      positiveDummyStr: String
+  ): Array[String] = {
+
+    val n = indices.size
+
+    // The length of the output array is n + 3.
+    //
+    // The first row is the shared features. These are features that are not label dependent.
+    // Then comes two dummy classes.  These are to make the probabilities work out.
+    // Then come the features for each of the n labels.
+    val x = new Array[String](n + 3)
+
+    val shared = VwRowCreator.unlabeledVwInput(features, defaultNs, namespaces, false)
+    x(0) = SharedFeatureIndicator + shared
+
+    // These string interpolations are computed over and over but will always be the same
+    // for a given dummyClassNs.
+    x(1) = negativeDummyStr
+    x(2) = positiveDummyStr
+
+    // This is mutable because we want speed.
+    var i = 0
+    while (i < n) {
+      val labelInd = indices(i)
+
+      // TODO or positives.contains(labelInd)?
+      val dv = if (positiveLabelIndices(i)) PositiveCost else NegativeCost
+      x(i + 3) = s"$labelInd:$dv |$classNs _$labelInd"
+      i += 1
+    }
+
+    x
+  }
+
+  /**
+    * Produce a multi-line input to be consumed by the underlying ''CSOAA LDF'' VW model.
+    * @param features (non-label dependent) features shared across all labels.
+    * @param indices the indices `labels` into the sequence of all labels encountered
+    *                during training.
+    * @param defaultNs the indices into `features` that should be placed in VW's default
+    *                  namespace.
+    * @param namespaces the indices into `features` that should be associated with each
+    *                   namespace.
+    * @param classNs a namespace for features associated with class labels
+    * @return an array to be passed directly to an underlying `VWActionScoresLearner`.
+    */
+  private[aloha] def predictionInput(
+      features: IndexedSeq[Sparse],
+      indices: sci.IndexedSeq[Int],
+      defaultNs: List[Int],
+      namespaces: List[(String, List[Int])],
+      classNs: String
+  ): Array[String] = {
+
+    val n = indices.size
+
+    // Use a (mutable) array (and iteration) for speed.
+    // The first row is the shared features. These are features that are not label dependent.
+    // Then come the features for each of the n labels.
+    val x = new Array[String](n + 1)
+
+    val shared = VwRowCreator.unlabeledVwInput(features, defaultNs, namespaces, false)
```

Comment:
I see that this called function is tail recursive with the following comment from @deaktator 

```
// RMD 2015-06-12: GOD I HATE THIS CODE!!!  Maybe functionalize it in the future!
```

Since we're concerned about performance here, is this worth revisiting now?

Reply:
This is more performant than an immutable implementation.  That comment was in reference to mutability, not performance.